### DISCUSSION TITLE
Editing AST

### DISCUSSION THREAD

**Discussion by vjpr** [tree-sitter/tree-sitter#2553] [Q&A] [answered]
I want to add a comment to the top of every function I find. When the user saves the file, or even as they finish typing a function.

In:

```
function foo() {}

function bar() {}
```

Out:

```
// foo
function foo() {}

// bar
function bar() {}
```

1. Is AST editing a use case for tree-sitter? From what I read, it seems not.

What is the point of `tree.edit`. I'm guessing its to track text editor edits (single letter, cut/copy/paste).

Can I run `tree.edit` multiple times before reparsing. Does that make sense?

2. Say I only want to only do this "comment" use case above - nothing too fancy, will using `tree.edit` be more performant than say using the `typescript` compiler api (maybe with language service) to parse to the full AST, make change, and write back?

Would tree-sitter give me more performance than typescript compiler api and language service? I guess a big advantage is that it can support multiple grammars.

But it seems like AST editing should be done with a language-specific tool.

**Comment by ahlinc** [comment]
> Is AST editing a use case for tree-sitter? From what I read, it seems not.

It isn't but nothing stops to use Tree-sitter in a way for what it was also built by calculating expected positions for edits by using initial Tree-sitter's tree after parsing and insert all edits as text right into an input text and in parallel inform Tree-sitter's initial tree about such edits by using `tree.edit()` method. All such edits needs be applied in a bottom up order to don't do offsets recalculation due to moved strings because of previous upper edits. If there is a need to get a new _edited AST_ then it's enough to call Tree-sitter parser again on the edited input and additionally pass an edited initial tree as an old tree _(parsing will happen much faster)_.

> What is the point of tree.edit. I'm guessing its to track text editor edits (single letter, cut/copy/paste).

I've described above.

> Can I run tree.edit multiple times before reparsing. Does that make sense?

Yes, it makes sense.

> will using tree.edit be more performant than say using the typescript compiler api (maybe with language service) to parse to the full AST, make change, and write back?
> Would tree-sitter give me more performance than typescript compiler api and language service?

I don't know and didn't see such comparisons, if you'd measure it that would be awesome.

> But it seems like AST editing should be done with a language-specific tool.

How do you imagine to craft a very complex edits with AST modifications? It's a ton of work that would be hard to maintain later due to it's relativity and an indirection in representation. Even in Rust where code generation is very powerful, a lot of this process based on thing that many modifications applied as snippets of code _(`paste` crate)_ _(translated to AST pieces but that's just a technical detail)_ because it's easier to figure out of what's going on in a code instead of looking on to an AST modification expressed in terms of data structures that represent an AST.

**Reply by dgutson** [reply]
What about doing a call expansion? For example I need to replace a function call with the body of the function defined previously. What's the best way of doing it?

**Comment by jeff-hykin** [comment, marked as answer]
The existing answer(s) also didn't really make sense to me. But I think I can explain.


## How to edit a tree (refactor/formatter)

*TLDR:* Skip to "helper function" below

I was looking for an api to replace stuff, like this:

```js
node.replaceWith("blah blah blah")
console.log(tree.rootNode.text)
// ^gives blah blah blah instead of whatever the node used to be
```

There is an efficent way to do this, but we've got to roll up our sleeves just a little.

### 1. Pass a **function** into the parser rather than a string

```js
import Parser  from "tree-sitter"
import javaScript from "tree-sitter-javascript"

const parser = new Parser()
parser.setLanguage(javaScript)

let sourceCode = 'console.log("howdy")'
let tree = parser.parse(()=>sourceCode) // here
```

### 2. Call the .edit on the tree as many times as you need

```js
// for example: node.replaceWith("blah blah blah")
let node = tree.rootNode.children[0].children[0].children[1].children[1]
let replacementText = `"hello Jeff"`

// get the node position info
const {
    startPosition: originalStart,
    endPosition: originalEnd,
    startIndex: originalStartIndex,
    endIndex: originalEndIndex 
} = node

// compute what the new row-column will be
const newNumberOflines = replacementText.match(/\n/g)?.length || 0
let newEndColumn = originalStart.column
if (newNumberOflines == 0) {
    newEndColumn = originalStart.row + replacementText.length
} else {
    newEndColumn = replacementText.split("\n").slice(-1)[0].length
}

// (this updates all the indices of all the nodes)
tree.edit({
    startIndex: originalStartIndex,
    oldEndIndex: originalEndIndex,
    newEndIndex: originalStartIndex + replacementText.length,
    startPosition: originalStart,
    oldEndPosition: originalEnd,
    newEndPosition: { row: originalStart.row + newNumberOflines, column: newEndColumn,  } 
})
```

### 3. Mutate the root string manually

```js
sourceCode = tree.string.slice(0, originalStartIndex) + replacementText + tree.string.slice(originalEndIndex) 
``` 

### Outcome

```js
console.log(tree.rootNode.text)
```

Prints:
```
console.log("hello Jeff")
```


## Helper function

Here's that process but wrapped up into a function, 

```js
// usage example:
let sourceCode = 'console.log("howdy")'
let tree = parser.parse(()=>sourceCode)
sourceCode = replaceNodeOn(tree, node, sourceCode, `"hello Jeff"`)

// now stuff like the following works:
// tree.rootNode.children[0].text

function replaceNodeOn(tree, node, sourceCode, replacement) {
  // get the node position info
  const {
    startPosition: originalStart,
    endPosition: originalEnd,
    startIndex: originalStartIndex,
    endIndex: originalEndIndex,
  } = node

  // compute what the new row-column will be
  const newNumberOflines = replacement.match(/\n/g)?.length || 0
  let newEndCol = originalStart.column;
  if (newNumberOflines == 0) {
    newEndCol = originalStart.column + replacement.length
  } else {
    newEndCol = replacement.split("\n").slice(-1)[0].length
  }

  // updates all the indices of all the nodes
  tree.edit({
    startIndex: originalStartIndex,
    oldEndIndex: originalEndIndex,
    newEndIndex: originalStartIndex + replacement.length,
    startPosition: originalStart,
    oldEndPosition: originalEnd,
    newEndPosition: { row: originalStart.row + newNumberOflines, column: newEndCol },
  })

  return sourceCode.slice(0, originalStartIndex) + replacement + sourceCode.slice(originalEndIndex)
}
```


## Caveats

Theres basically two caveats. 

1. Replacement doesnt re-parse. You could replace a node with absolutely invalid/impossible text. All the other nodes merely have their indices updated. Which is fine, even desirable sometimes, but just FYI: not a re-parse.
2. If we do a full re-parse (feeding in the old tree to make the re-parse fast), then none of the node objects will be preserved. E.g. even if you only edit the last child node, `oldTree.childNode1` will not be the same object as `newTree.childNode1`. AFAIK there is no node-preserving way to do partial re-parsing. Here was an example I found in the tree-sitter codebase.

```js
async function handleCodeChange(editor, changes) {
    const newText = codeEditor.getValue() + "\n"
    const edits = tree && changes && changes.map(treeEditForEditorChange)

    const start = performance.now()
    if (edits) {
        for (const edit of edits) {
            tree.edit(edit)           // <- here
        }
    }
    const newTree = parser.parse(newText, tree) // NOTE: based on my tests, none of the node ID's are preserved 
    const duration = (performance.now() - start).toFixed(1)

    updateTimeSpan.innerText = `${duration} ms`
    if (tree) tree.delete()
    tree = newTree
    parseCount++
    renderTreeOnCodeChange()
    runTreeQueryOnChange()
    saveStateOnChange()
}
```

It would be cool if the node-id system used a pseudo random number generator that always had the same seed, but was incremented with every operation. Then at least the begining part of the tree could be preserved.

**Reply by geomaster** [reply]
If anybody stumbles on this (like I did), it appears that the code (https://github.com/tree-sitter/tree-sitter/discussions/2553#discussioncomment-9976343) mixes up rows and columns.

The correct code for `replaceNodeOn()` should be something like:

```typescript
function replaceNodeOn(tree: Tree, needle: SyntaxNode, haystack: string, replace: string) {
  // get the node position info
  const {
    startPosition: originalStart,
    endPosition: originalEnd,
    startIndex: originalStartIndex,
    endIndex: originalEndIndex,
  } = needle;

  // compute what the new row-column will be
  const newNumberOflines = replace.match(/\n/g)?.length || 0;
  let newEndCol = originalStart.column; // <-- column is actually the character in line
  if (newNumberOflines == 0) {
    newEndCol = originalStart.column + replace.length;
  } else {
    newEndCol = replace.split("\n").slice(-1)[0].length;
  }

  // updates all the indices of all the nodes
  tree.edit({
    startIndex: originalStartIndex,
    oldEndIndex: originalEndIndex,
    newEndIndex: originalStartIndex + replace.length,
    startPosition: originalStart,
    oldEndPosition: originalEnd,
    newEndPosition: { row: originalStart.row + newNumberOflines, column: newEndCol }, // <-- row is actually the line
  });

  return haystack.slice(0, originalStartIndex) + replace + haystack.slice(originalEndIndex);
}
```

**Reply by jeff-hykin** [reply]
Thanks for the fix @geomaster ! I went ahead an ported it back to the comment to try and save people the trouble.

**Reply by vladinator1000** [reply]
The fact that node object IDs are not preserved in the web runtime caught me off-guard. This image sold me on tree-sitter originally https://youtu.be/Jes3bD6P0To?t=104: 
<img width="1879" height="1051" alt="image" src="https://github.com/user-attachments/assets/6205ba77-1617-48e0-b8e2-a29ce013b407" />

In practice, if you want to use the edited tree to, let's say, re-render stuff in React, do you have to maintain your own stable IDs? 

I see there's an `node.id` which gives me some hope 👀 https://github.com/tree-sitter/tree-sitter/blob/13d4db8bb484064e5bfe0de22382c0e2bf391fd7/lib/binding_web/src/node.ts#L47-L59

Edit: Wrote a reproduction test showing unstable node ID https://github.com/tree-sitter/tree-sitter/pull/5039

**Comment by RedCMD** [comment]
> If a new tree is created based on an older tree, and a node from the old tree is reused in the process, then that node will have the same id in both trees. 

afaik this is not correct?
I don't remember ever seeing the same id across edits

there is `hasChanges` on nodes
tho in practice its not very useful(?)
as iirc it just marks the very nodes you marked in `.edit()`

---

currently the approach for keeping track of nodes between edits is to track the ranges of nodes and simply just guess

here's my code for reparsing an embedded language
(regex inside json)
https://github.com/RedCMD/TmLanguage-Syntax-Highlighter/blob/main/src/TreeSitter.ts#L376-L505

vim and other editors use the exact same approach
I remember being shown the code
but can't seem to find it
maybe @ObserverOfTime knows?
